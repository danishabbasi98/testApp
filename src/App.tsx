import React from 'react';
import './App.css';


function App() {
  return (
    <div id='app' className='app'>
      <h1>HelloApp</h1>
      
    </div>
  );
}

export default App;
