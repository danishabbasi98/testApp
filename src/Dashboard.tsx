import { Component ,useState ,useEffect } from "react";
import { Button, Card, CardBody, CardFooter, CardText, CardTitle, Col, Container, Navbar, NavbarBrand, Row } from "reactstrap";
import { 
    IoSchoolOutline, 
    IoMan,
    IoLayersOutline,
    IoWomanOutline, 
} from "react-icons/io5";
import axios from "axios";
import CreationModal from "./CreationModal";


interface MyState {
    students : [];
    teachers : [];
    subjects : [];
    isOpen : boolean;

}



function refreshPage(this: any) {
    window.location.reload();
    // this.setState((prevState: { isOpen: any; }) => ({ isOpen : !prevState.isOpen}));
  }


function RenderStudent({st}) {
    console.log("st : " + JSON.stringify(st))

    return (
       <Row>
            <Col sm = "12">
               <Card body style={{marginBottom : 10}}>
                   <CardTitle tag = "h5">
                       <IoMan className="font-size-xl"/> {st.name}
                   </CardTitle>
                   <CardBody >
                       <Row>
                           <Col sm = "4" className="text-center">
                               <span className="font-weight-bold">Name </span>
                               <span>{st.name}</span>
                           </Col>
                           <Col sm = "4" className="text-center">
                               <span className="font-weight-bold"> Subject </span>
                               <span> {st.name}</span>
                           </Col>
                           <Col sm = "4" className="text-center">
                               <span className="font-weight-bold"> Teacher </span>
                               <span> {st.name}</span>
                           </Col>
                       </Row>
                   </CardBody>
                   <CardFooter>
                       <Row>
                           <Col sm-6>
                               <Button block outline color="primary">
                                   Edit
                               </Button>
                           </Col>
                           <Col sm-6>
                               <Button data-testid = "Delete-button" block outline color="danger" onClick={() => deleteStudent(st.id)} 
                               onClickCapture={()=> refreshPage()}
                               >
                                   Delete
                               </Button>
                           </Col>
                       </Row>
                   </CardFooter>
               </Card>
           </Col> 
       </Row> 
    );
}


export default class Dashboard extends Component<{}, MyState> {
    state : MyState = {
        isOpen : false,
        students : [],
        teachers : [],
        subjects : [],
    };
    

    constructor(props) {
        super(props);
        // this.renderStudent = this.renderStudent.bind(this);
      }


    componentDidMount(){
        axios.get('http://localhost:8080/students')
        .then(res => {
            const students = res.data
            console.log("students : " + JSON.stringify(students))
            this.setState({students : students})
        })
        axios.get('http://localhost:8080/teachers')
        .then(res => {
            const teachers = res.data
            console.log("Teachers : " + JSON.stringify(teachers))
            this.setState({teachers : teachers})
        })
        axios.get('http://localhost:8080/subjects')
        .then(res => {
            const subjects = res.data
            console.log("Subjects : " + JSON.stringify(subjects))
            this.setState({subjects : subjects})
        })
    }
    
    toggle = () => {
        this.componentDidMount();
        this.setState((prevState) => ({ isOpen : !prevState.isOpen}));
    }

    render(){
        
        return(
            <div>
                <Navbar color="dark" light mb-2>
                    <NavbarBrand className="text-white">
                        <IoSchoolOutline className = "font-size-xxl"/>
                        <span data-testid  = "span-name" className="font-size-l ml-3">School Manager Application</span>

                    </NavbarBrand>
                    
                </Navbar>

                <Container className="mt-3">
                    <Row>
                        <Col sm ='4'>
                            <Card body >
                                <CardTitle tag = "h5">
                                    <IoMan className="font-size-xl"/> {this.state.students.length} Students
                                </CardTitle>
                                <CardText>
                                    With supportking text below as a natural lead in to additional content.
                                </CardText>
                                <Button data-testid = "button-test">Manage Students</Button>
                            </Card>

                        </Col>
                        <Col sm ='4'>

                        <Card body >
                                <CardTitle tag = "h5">
                                    <IoSchoolOutline className="font-size-xl"/> {this.state.teachers.length} Teachers
                                </CardTitle>
                                <CardText>
                                    With supportking text below as a natural lead in to additional content.
                                </CardText>
                                <Button>Manage Teacher</Button>
                            </Card>
                        </Col>
                        <Col sm ='4'>

                        <Card body >
                                <CardTitle tag = "h5">
                                    <IoLayersOutline className="font-size-xl"/> {this.state.subjects.length}  Subjects
                                </CardTitle>
                                <CardText>
                                    With supportking text below as a natural lead in to additional content.
                                </CardText>
                                <Button>Manage Subjects</Button>
                            </Card>
                        </Col>
                    </Row>

                </Container>
                <CreationModal isOpen={this.state.isOpen} toggle={this.toggle}></CreationModal>
                <Container className="mt-4">
                    <Row>
                        <Col sm ='12'>        
                            <Button data-testid = "button" block color ="success" onClick={this.toggle}>
                                <span className="font-size-l">Create New Student </span>
                            </Button>       
                        </Col>    
                    </Row>
                </Container>

                <Container className="mt-4">
                    <h2 title="List Student">All Students List</h2>
                    { this.state.students.length > 0 && this.state.students.map(student => <RenderStudent st={student} />)}
                    {this.toggle}

                </Container>
            </div>
        );
    }



}
function deleteStudent(id: any): void {
    axios.delete(`http://localhost:8080/students/delete/${id}`)
    .then(res => {
        console.log("Delete student : ID :: " + JSON.stringify(id))
        // refreshPage();
    }).catch(Error);
    console.log("ERROR : ID :: " + JSON.stringify(id))
}
