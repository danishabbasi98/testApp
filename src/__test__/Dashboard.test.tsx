
import react from "react";
import ReactDOM from "react-dom";
import Button from "../Dashboard";
import "@testing-library/jest-dom/extend-expect";
import renderer from "react-test-renderer";
import Dashboard from "../Dashboard";
import App from "../App";
import { request  } from "http";
import {rest} from 'msw'
import {setupServer} from 'msw/node'
import {render, cleanup, fireEvent, waitFor, screen} from '@testing-library/react'
import '@testing-library/jest-dom'


afterEach(cleanup)
test("render without crashing" , ()=> {
    const root = document.createElement("div");
    // root.setAttribute("class", "app")
    ReactDOM.render(<App/>, root);
    expect(root.querySelector("div")?.textContent).toBe("HelloApp");
    // expect(root.querySelector("span")?.textContent).toBe("School Manager Application");
});
// describe("Api called ", () => {
//   test("fetch the data", async () => {
//     const result = Dashboard
//     console.log("DashBoard"+Dashboard);
//     // expect(result).toBeInTheDocument()
//   })
// });

test("button working" , ()=> {
    // const root = document.createElement("div");
    // root.setAttribute("class", "app")
    // ReactDOM.render(<App/>, root);
    const { getByTestId } = render(<Button />);
    expect(getByTestId("button-test")).toHaveTextContent("Manage Students");
    
});

test("all buttons" , ()=> {
    const { getByTestId, getByText, getByLabelText, getAllByTitle} = render(<Dashboard/>);

    // getByLabelText("Create New Student");
    getByText("Create New Student");
    getByTestId("button-test");
    getByTestId("span-name");
    getAllByTitle("List Student");
  
});

test("allows users to add items to their list", () =>{
    const { getByText, getByLabelText } = render(<Dashboard/>);

    const input = getByText("Create New Student");
    // fireEvent.change(input, { target:{ value: "Create a Student"}});
    fireEvent.click(getByText("Create New Student"));
});

describe("Api called ", () => {
    test("fetch the data", async () => {
      const { findByText } = render(<Dashboard/>)
      expect(await findByText('School Manager Application')).toBeInTheDocument()
    })
  });

// describe("GET / ", () => {
//     // we will use supertest to test HTTP requests/responses
//     // const request = require("supertest");
//     // we also need our app for the correct routes!
//     const app = require("../Dashboard");
//     test("It should respond with an array of students", async () => {
//       const response = await request(app).get("http://localhost:8080/subjects");
//     //   expect(response.body).toEqual(["Elie", "Matt", "Joel", "Michael"]);
//       expect(response.statusCode).toBe(200);
//     });
//   });



  // const server = setupServer(
  //   rest.get('/greeting', (req, res, ctx) => {
  //     return res(ctx.json({greeting: 'hello there'}))
  //   }),
  // )
  
  // beforeAll(() => server.listen())
  // afterEach(() => server.resetHandlers())
  // afterAll(() => server.close())
  
  // test('loads and displays greeting', async () => {
  //   render(<App url="http://localhost:8080/subjects" />)
  
  //   fireEvent.click(screen.getByText('Load Greeting'))
  
  //   await waitFor(() => screen.getByRole('heading'))
  
  //   expect(screen.getByRole('heading')).toHaveTextContent('hello there')
  //   expect(screen.getByRole('button')).toBeDisabled()
  // })


// test("allows student to delete", () =>{
//     const { getByText, getByTestId } = render(<Dashboard/>);

//     const input = getByText("Delete");
//     // fireEvent.change(input, { target:{ value: "Create a Student"}});
//     fireEvent.click(getByText("Delete"));
// });

// test("should render component" , ()=> {
//     render(<Dashboard/>);
//     const dashElement = document.createElement("div");
//     ReactDOM.render(<Button></Button>, dashElement);

// });

// it("render button correctly ", ()=> {
//     const { getByTestId } = render(<Button />);
//     expect(getByTestId("button")).toHaveTextContent("Create New Student");
// });

// it("render button correctly ", ()=> {
//     const { getByTestId } = render(<Button label="save"/>);
//     expect(getByTestId("button")).toHaveTextContent("save");
// });

it("matches snapshot 1", ()=> {
    const tree = renderer.create(<Button ></Button>).toJSON();
    expect(tree).toMatchSnapshot();
});

it("matches snapshot 2", ()=> {
    const tree = renderer.create(<Button ></Button>).toJSON();
    expect(tree).toMatchSnapshot();
});