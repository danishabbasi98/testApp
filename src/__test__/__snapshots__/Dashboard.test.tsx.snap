// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1 1`] = `
<div>
  <nav
    className="navbar navbar-light bg-dark"
    mb-2={true}
  >
    <div
      className="container-fluid"
    >
      <a
        className="text-white navbar-brand"
      >
        <svg
          className="font-size-xxl"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M32 192L256 64l224 128-224 128L32 192z"
            fill="none"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="32"
          />
          <path
            d="M112 240v128l144 80 144-80V240m80 128V192M256 320v128"
            fill="none"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="32"
          />
        </svg>
        <span
          className="font-size-l ml-3"
          data-testid="span-name"
        >
          School Manager Application
        </span>
      </a>
    </div>
  </nav>
  <div
    className="mt-3 container"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-4"
      >
        <div
          className="card card-body"
        >
          <h5
            className="card-title"
          >
            <svg
              className="font-size-xl"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="256"
                cy="56"
                r="56"
              />
              <path
                d="M304 128h-96a64.19 64.19 0 00-64 64v107.52c0 10.85 8.43 20.08 19.27 20.47A20 20 0 00184 300v-99.73a8.18 8.18 0 017.47-8.25 8 8 0 018.53 8V489a23 23 0 0023 23 23 23 0 0023-23V346.34a10.24 10.24 0 019.33-10.34A10 10 0 01266 346v143a23 23 0 0023 23 23 23 0 0023-23V200.27a8.18 8.18 0 017.47-8.25 8 8 0 018.53 8v99.52c0 10.85 8.43 20.08 19.27 20.47A20 20 0 00368 300V192a64.19 64.19 0 00-64-64z"
              />
            </svg>
             
            0
             Students
          </h5>
          <p
            className="card-text"
          >
            With supportking text below as a natural lead in to additional content.
          </p>
          <button
            aria-label={null}
            className="btn btn-secondary"
            data-testid="button-test"
            onClick={[Function]}
          >
            Manage Students
          </button>
        </div>
      </div>
      <div
        className="col-sm-4"
      >
        <div
          className="card card-body"
        >
          <h5
            className="card-title"
          >
            <svg
              className="font-size-xl"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M32 192L256 64l224 128-224 128L32 192z"
                fill="none"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="32"
              />
              <path
                d="M112 240v128l144 80 144-80V240m80 128V192M256 320v128"
                fill="none"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="32"
              />
            </svg>
             
            0
             Teachers
          </h5>
          <p
            className="card-text"
          >
            With supportking text below as a natural lead in to additional content.
          </p>
          <button
            aria-label={null}
            className="btn btn-secondary"
            onClick={[Function]}
          >
            Manage Teacher
          </button>
        </div>
      </div>
      <div
        className="col-sm-4"
      >
        <div
          className="card card-body"
        >
          <h5
            className="card-title"
          >
            <svg
              className="font-size-xl"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M434.8 137.65l-149.36-68.1c-16.19-7.4-42.69-7.4-58.88 0L77.3 137.65c-17.6 8-17.6 21.09 0 29.09l148 67.5c16.89 7.7 44.69 7.7 61.58 0l148-67.5c17.52-8 17.52-21.1-.08-29.09zM160 308.52l-82.7 37.11c-17.6 8-17.6 21.1 0 29.1l148 67.5c16.89 7.69 44.69 7.69 61.58 0l148-67.5c17.6-8 17.6-21.1 0-29.1l-79.94-38.47"
                fill="none"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="32"
              />
              <path
                d="M160 204.48l-82.8 37.16c-17.6 8-17.6 21.1 0 29.1l148 67.49c16.89 7.7 44.69 7.7 61.58 0l148-67.49c17.7-8 17.7-21.1.1-29.1L352 204.48"
                fill="none"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="32"
              />
            </svg>
             
            0
              Subjects
          </h5>
          <p
            className="card-text"
          >
            With supportking text below as a natural lead in to additional content.
          </p>
          <button
            aria-label={null}
            className="btn btn-secondary"
            onClick={[Function]}
          >
            Manage Subjects
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="mt-4 container"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-12"
      >
        <button
          aria-label={null}
          className="btn btn-success d-block w-100"
          data-testid="button"
          onClick={[Function]}
          type="button"
        >
          <span
            className="font-size-l"
          >
            Create New Student 
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    className="mt-4 container"
  >
    <h2
      title="List Student"
    >
      All Students List
    </h2>
  </div>
</div>
`;

exports[`matches snapshot 2 1`] = `
<div>
  <nav
    className="navbar navbar-light bg-dark"
    mb-2={true}
  >
    <div
      className="container-fluid"
    >
      <a
        className="text-white navbar-brand"
      >
        <svg
          className="font-size-xxl"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          strokeWidth="0"
          style={
            Object {
              "color": undefined,
            }
          }
          viewBox="0 0 512 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M32 192L256 64l224 128-224 128L32 192z"
            fill="none"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="32"
          />
          <path
            d="M112 240v128l144 80 144-80V240m80 128V192M256 320v128"
            fill="none"
            strokeLinecap="round"
            strokeLinejoin="round"
            strokeWidth="32"
          />
        </svg>
        <span
          className="font-size-l ml-3"
          data-testid="span-name"
        >
          School Manager Application
        </span>
      </a>
    </div>
  </nav>
  <div
    className="mt-3 container"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-4"
      >
        <div
          className="card card-body"
        >
          <h5
            className="card-title"
          >
            <svg
              className="font-size-xl"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <circle
                cx="256"
                cy="56"
                r="56"
              />
              <path
                d="M304 128h-96a64.19 64.19 0 00-64 64v107.52c0 10.85 8.43 20.08 19.27 20.47A20 20 0 00184 300v-99.73a8.18 8.18 0 017.47-8.25 8 8 0 018.53 8V489a23 23 0 0023 23 23 23 0 0023-23V346.34a10.24 10.24 0 019.33-10.34A10 10 0 01266 346v143a23 23 0 0023 23 23 23 0 0023-23V200.27a8.18 8.18 0 017.47-8.25 8 8 0 018.53 8v99.52c0 10.85 8.43 20.08 19.27 20.47A20 20 0 00368 300V192a64.19 64.19 0 00-64-64z"
              />
            </svg>
             
            0
             Students
          </h5>
          <p
            className="card-text"
          >
            With supportking text below as a natural lead in to additional content.
          </p>
          <button
            aria-label={null}
            className="btn btn-secondary"
            data-testid="button-test"
            onClick={[Function]}
          >
            Manage Students
          </button>
        </div>
      </div>
      <div
        className="col-sm-4"
      >
        <div
          className="card card-body"
        >
          <h5
            className="card-title"
          >
            <svg
              className="font-size-xl"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M32 192L256 64l224 128-224 128L32 192z"
                fill="none"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="32"
              />
              <path
                d="M112 240v128l144 80 144-80V240m80 128V192M256 320v128"
                fill="none"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="32"
              />
            </svg>
             
            0
             Teachers
          </h5>
          <p
            className="card-text"
          >
            With supportking text below as a natural lead in to additional content.
          </p>
          <button
            aria-label={null}
            className="btn btn-secondary"
            onClick={[Function]}
          >
            Manage Teacher
          </button>
        </div>
      </div>
      <div
        className="col-sm-4"
      >
        <div
          className="card card-body"
        >
          <h5
            className="card-title"
          >
            <svg
              className="font-size-xl"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M434.8 137.65l-149.36-68.1c-16.19-7.4-42.69-7.4-58.88 0L77.3 137.65c-17.6 8-17.6 21.09 0 29.09l148 67.5c16.89 7.7 44.69 7.7 61.58 0l148-67.5c17.52-8 17.52-21.1-.08-29.09zM160 308.52l-82.7 37.11c-17.6 8-17.6 21.1 0 29.1l148 67.5c16.89 7.69 44.69 7.69 61.58 0l148-67.5c17.6-8 17.6-21.1 0-29.1l-79.94-38.47"
                fill="none"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="32"
              />
              <path
                d="M160 204.48l-82.8 37.16c-17.6 8-17.6 21.1 0 29.1l148 67.49c16.89 7.7 44.69 7.7 61.58 0l148-67.49c17.7-8 17.7-21.1.1-29.1L352 204.48"
                fill="none"
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth="32"
              />
            </svg>
             
            0
              Subjects
          </h5>
          <p
            className="card-text"
          >
            With supportking text below as a natural lead in to additional content.
          </p>
          <button
            aria-label={null}
            className="btn btn-secondary"
            onClick={[Function]}
          >
            Manage Subjects
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    className="mt-4 container"
  >
    <div
      className="row"
    >
      <div
        className="col-sm-12"
      >
        <button
          aria-label={null}
          className="btn btn-success d-block w-100"
          data-testid="button"
          onClick={[Function]}
          type="button"
        >
          <span
            className="font-size-l"
          >
            Create New Student 
          </span>
        </button>
      </div>
    </div>
  </div>
  <div
    className="mt-4 container"
  >
    <h2
      title="List Student"
    >
      All Students List
    </h2>
  </div>
</div>
`;
